fix(keymap): backport use of correct `derivation_path` for keys with origin

oleonardolima opened the pull request:
backports #872

### Description

It backports the fix for deriving the private key correctly in scenarios where the key origin is available, see #872.

The GetKey for DescriptorSecretKey (and thus KeyMap) returned the wrong key when an extended private key with key origin info was queried via KeyRequest::Bip32.

The key request path is master-relative, but the xkey is anchored at its origin (e.g. [d34db33f/84h/1h/0h]). The old code stripped the path returned by matches(); the fix strips the origin prefix instead (and uses the full path when there is no origin) before deriving from the xkey.

cherry-picks 20ae466ec598f7cf973027cefbb433a46b3a59cf and a2f4497598c9b388d22e5a4dececd99896774116.

### Changelog notice

```
### Fixed
- keymap: fix `GetKey` derivation path for `Xprv` with key origin info [#872](https://github.com/rust-bitcoin/rust-miniscript/pull/872)
```

### Checklists

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

tcharding commented:
Do you want review mate? Why's it draft?

oleonardolima commented:
> Do you want review mate? Why's it draft?

Yeah, I was just waiting for CI, it should be ready to go.

tcharding reviewed (approved):
ACK 10fd701d3c316a67c65b050f2e8aa50bf0bf11c6

oleonardolima committed:
fix(key_map): derivation_path for `Xprv` with key origin info

- backports #872 fixes for the implementation of `GetKey` for `Xprv` with
  `KeyRequest::Bip32`, when the the key_origin information is available.

notmandatory commented:
@oleonardolima just a ping, can this be updated and made ready to review now?

oleonardolima commented:
> @oleonardolima just a ping, can this be updated and made ready to review now?

I'll rebase it and get it ready for review today.

oleonardolima committed:
test(key_map): improve `Xprv` key origin tests

- backports introduced #872 table-driven test cases for
  `get_key_xpriv_with_key_origin` with named fields
- drop redundant key parsing, and build requests from explicit
  key sources

Assisted-by: Claude Opus 4.8

oleonardolima commented:
> @oleonardolima just a ping, can this be updated and made ready to review now?

@notmandatory I updated the original fix in #872 (also needs review), I'm backporting it here now.

oleonardolima commented:
> > @oleonardolima just a ping, can this be updated and made ready to review now?
> 
> @notmandatory I updated the original fix in #872 (also needs review), I'm backporting it here now.

I updated this PR and added new test scenarios, it should be ready to review.

cc'in @noahjoeris as you had interest in this and a review would be appreciated.

notmandatory reviewed (approved):
ACK 3f777676ce36581b6e3a5ecb1ddad6630db03dd5

Accurately back ports #872, should be good to go after that one is merged.

noahjoeris reviewed (approved):
ACK 3f77767

oleonardolima commented:
ping @tcharding @apoelstra for awareness.

oleonardolima commented:
@apoelstra thanks for the quick review in #872, let me know if I need to do any update here in the backport.

apoelstra commented:
Sorry, the problem is on my end. My local CI is not passing on the 12.x branch and I need to investigate why.

apoelstra commented:
Oh, no, the problem is on your end :). In 122105d4cce582ca70a1690abfa41de22df03a6d your original test does not compile:

```
error[E0599]: no method named `get_key` found for struct `std::collections::BTreeMap<K, V, A>` in the current scope
   --> src/descriptor/key_map.rs:316:14
    |
315 |           let pk = keymap
    |  __________________-
316 | |             .get_key(key_request, &secp)
    | |_____________-^^^^^^^
    |
help: there is a method `get_key_value` with a similar name, but with different arguments
   --> /rustc/ac68faa20c58cbccd01ee7208bf3b6e93a7d7f96/library/alloc/src/collections/btree/map.rs:783:4
    = note: the full name for the type has been written to 'target/lib/miniscript-779a623562.long-type-17215175838718595439.txt'
    = note: consider using `--verbose` to print the full type name to the console
```

In this project we require that all commits compile and pass unit tests.

Sorry again for the long delay on my part! I had some weird NixOS issues all week and had assumed that the failure here was part of that.

oleonardolima commented:
> Oh, no, the problem is on your end :). In [122105d](https://github.com/rust-bitcoin/rust-miniscript/commit/122105d4cce582ca70a1690abfa41de22df03a6d) your original test does not compile:
> 
> ```
> error[E0599]: no method named `get_key` found for struct `std::collections::BTreeMap<K, V, A>` in the current scope
>    --> src/descriptor/key_map.rs:316:14
>     |
> 315 |           let pk = keymap
>     |  __________________-
> 316 | |             .get_key(key_request, &secp)
>     | |_____________-^^^^^^^
>     |
> help: there is a method `get_key_value` with a similar name, but with different arguments
>    --> /rustc/ac68faa20c58cbccd01ee7208bf3b6e93a7d7f96/library/alloc/src/collections/btree/map.rs:783:4
>     = note: the full name for the type has been written to 'target/lib/miniscript-779a623562.long-type-17215175838718595439.txt'
>     = note: consider using `--verbose` to print the full type name to the console
> ```
> 
> In this project we require that all commits compile and pass unit tests.
> 
> Sorry again for the long delay on my part! I had some weird NixOS issues all week and had assumed that the failure here was part of that.

Oh, sorry for the oversight!

I updated the first commit 10537996c385405c3b373209214fdd0709662f90 to drop the broken and newly added test as it was being refactored in the second one d01ee72.

Let me know if there's anything else to address.

apoelstra reviewed (approved):
ACK d01ee729342bfa84f1fa0878410bd480582bb8a5; successfully ran local tests
